pywbem_mock/_wbemconnection_mock.py

```diff
@@ -0,0 +1,1571 @@
+#!/usr/bin/env python
+
+# (C) Copyright 2017 IBM Corp.
+# (C) Copyright 2017 Inova Development Inc.
+# All Rights Reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Mock of WBEMConnection to enable tests of pywbem without a WBEM server.
+
+This class mocks all of the pywbem WBEMConnectionoperations that communicate
+with a wbem server and provides the means to build a local CIMClass,
+CIMInstance, and CIMQualifierDeclaration repository that each of the
+reactor methods (i.e. the methods of WBEMConnection that simulate a wbem
+server use as the source of data for responses and to temporarily save
+CIMClasses, CIMQualifierDeclarations, and CIMInstances created by the
+WBEMConnection methods that create new objects (e.x. CreateInstance).
+
+It includes methods to add CIM objects to that repository.
+
+  - add_cimobjects adds python classes CIMClass, CIMInstance, and
+CIMQualifierDeclaration to the repository.
+
+  - compile_mof_str adds objects to the repository from mof definitions
+  contained in strings.  For classes to compile correctly, the qualifier
+  declarations for the target classes must exist in the repository.  For
+  instances to compile correctly, the corresponding classes must be in the
+  repository.
+
+  - compile_mof adds CIM objects to the repository from mof definitions
+  contained in mof files. For classes to compile correctly, the qualifier
+  declarations for the target classes must exist in the repository.  For
+  instances to compile correctly, the corresponding classes must be in the
+  repository.
+
+FakeWBEMConnection provides responses for the WBEMConnection methods based on
+the data in the repository including responding with the correct CIMError
+exceptions when error conditions are found.
+
+The repository allows use of multiple namespaces with the default being the
+standard pywmbe default namespace.
+
+This mocker uses the same APIs as the standard WBEMConnections (it actually
+uses the methods and only mocks the communication with the WBEMServer) and
+returns syntatically the same forms of data and exceptions.
+
+- Class methods:
+
+The methods that get data require the expected classes to be in the repository
+before the mock call. The methods are:
```

Comment:
And their superclasses. And CreateClass is not one that gets data but still requires its superclasses in the repo, right?

It seems to me that instead of making statements for these groups, we should make a general statement similar to the one at the end, that basically says:
* Any CIM objects expected to be returned by the methods need to be in the repository.
* Any creation methods put their input objects into the repository.
* Classes in the repo require their superclasses and referenced classes and embedded classes to also be in the repo.
* Instances in the repo require their creation classes also be in the repo.
* Qualifier values on objects in the repo require their qualifier declarations also to be in the repo.

Reply:
Actually we don't for some of the methods.   I am going to leave this one open to review for now.
It may be easier to actually require these limitations since it is fairly easy to generate the information with the compiler (messy with the add methods).

**KS: DISCUSSION**

Reply:
Now shown in list of discussion points, below.